File: app/src/main/java/com/example/waveplayer/random_playlist/ProbMap.java
package com.example.waveplayer.random_playlist;

import java.io.Serializable;
import java.util.HashMap;
import java.util.Random;

public class ProbMap<T> implements Serializable {

    private final HashMap<T, Double> probHashMap = new HashMap<>();

    public void put(T t, Double aDouble) {
        probHashMap.put(t, aDouble);
    }

    public boolean outcome(T t, Random random) {
        double prob = probHashMap.get(t);
        double randomChoice = random.nextDouble();
        return !(randomChoice > prob);
    }

    public boolean bad(T t, double percent) {
        if (percent >= 1.0 || percent <= 0.0) {
            throw new IllegalArgumentException(
                    "percent passed to bad() is not between 0.0 and 1.0 (exclusive)");
        }
        Double dub = probHashMap.get(t);
        if(dub == null){
            return false;
        }
        double prob = dub;
        boolean globalBad = true;
        for (Double d : probHashMap.values()) {
            if (prob <= d) {
                globalBad = false;
                break;
            }
        }
        double sub = (prob * percent);
            probHashMap.put(t, (prob - sub));
        return globalBad;
    }

    public boolean good(T t, double percent) {
        if (percent >= 1.0 || percent <= 0.0) {
            throw new IllegalArgumentException(
                    "percent passed to good() is not between 0.0 and 1.0 (exclusive)");
        }
        Double dub = probHashMap.get(t);
        if(dub == null){
            return false;
        }
        double prob = dub;
        boolean globalGood = true;
        for (Double d : probHashMap.values()) {
            if (prob >= d) {
                globalGood = false;
                break;
            }
        }
        double add = (prob * percent);
        probHashMap.put(t, Math.min(prob + add, 1.0));
        return globalGood;
    }

    public void clearProbabilities(){
        for(T t : probHashMap.keySet()){
            probHashMap.put(t, 1.0);
        }
    }

}

File: app/src/main/java/com/example/waveplayer/random_playlist/ProbMap.kt
package com.example.waveplayer.random_playlist

import java.io.Serializable
import java.util.*

class ProbMap<T> : Serializable {
    private val probHashMap: HashMap<T?, Double?>? = HashMap()
    fun put(t: T?, aDouble: Double?) {
        probHashMap[t] = aDouble
    }

    fun outcome(t: T?, random: Random?): Boolean {
        val prob: Double = probHashMap.get(t)
        val randomChoice = random.nextDouble()
        return randomChoice <= prob
    }

    fun bad(t: T?, percent: Double): Boolean {
        require(!(percent >= 1.0 || percent <= 0.0)) { "percent passed to bad() is not between 0.0 and 1.0 (exclusive)" }
        val dub = probHashMap.get(t) ?: return false
        var globalBad = true
        for (d in probHashMap.values) {
            if (dub <= d) {
                globalBad = false
                break
            }
        }
        val sub = dub * percent
        probHashMap[t] = dub - sub
        return globalBad
    }

    fun good(t: T?, percent: Double): Boolean {
        require(!(percent >= 1.0 || percent <= 0.0)) { "percent passed to good() is not between 0.0 and 1.0 (exclusive)" }
        val dub = probHashMap.get(t) ?: return false
        var globalGood = true
        for (d in probHashMap.values) {
            if (dub >= d) {
                globalGood = false
                break
            }
        }
        val add = dub * percent
        probHashMap[t] = Math.min(dub + add, 1.0)
        return globalGood
    }

    fun clearProbabilities() {
        for (t in probHashMap.keys) {
            probHashMap[t] = 1.0
        }
    }
}